

<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      <link rel="icon" href="../../assets/images/SIB_logo.svg">
      <meta name="generator" content="mkdocs-1.4.1, mkdocs-material-8.5.7">
    
    
      
        <title>Quality control - NGS - Quality control, Alignment, Visualisation</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.20d9efc8.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.cbb835fc.min.css">
        
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-YX61LNKWGY"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-YX61LNKWGY');
</script>

    
    
  <meta property="og:url"                content="https://sib-swiss.github.io/NGS-introduction-training/latest/" />
  <meta property="og:title"              content="NGS - Quality Control, Alignment, Visualisation" />
  <meta property="og:description"        content="Course on the introduction to next generation sequencing analysis" />
  <meta property="og:image"              content="https://sib-swiss.github.io/NGS-introduction-training/2022.11/assets/images/alignment_infograph.png" />
  
  <meta name="twitter:card" content="summary_large_image" />
  <meta name="twitter:title" content="NGS - Quality Control, Alignment, Visualisation" />
  <meta name="twitter:description" content="Course on the introduction to next generation sequencing analysis" />
  <meta name="twitter:image" content="https://sib-swiss.github.io/NGS-introduction-training/2022.11/assets/images/alignment_infograph.png" />

  
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="none" data-md-color-accent="none">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#learning-outcomes" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <div data-md-component="outdated" hidden>
        
      </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="NGS - Quality control, Alignment, Visualisation" class="md-header__button md-logo" aria-label="NGS - Quality control, Alignment, Visualisation" data-md-component="logo">
      
  <img src="../../assets/images/SIB_logo.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            NGS - Quality control, Alignment, Visualisation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Quality control
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/sib-swiss/NGS-introduction-training" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/NGS-introduction-training
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="NGS - Quality control, Alignment, Visualisation" class="md-nav__button md-logo" aria-label="NGS - Quality control, Alignment, Visualisation" data-md-component="logo">
      
  <img src="../../assets/images/SIB_logo.svg" alt="logo">

    </a>
    NGS - Quality control, Alignment, Visualisation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/sib-swiss/NGS-introduction-training" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/NGS-introduction-training
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../precourse/" class="md-nav__link">
        Precourse preparations
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../course_schedule/" class="md-nav__link">
        Course schedule
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4" type="checkbox" id="__nav_4" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_4">
          Day 1
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Day 1" data-md-level="1">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          Day 1
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../intro/" class="md-nav__link">
        Introduction
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../sequencing_technologies/" class="md-nav__link">
        Sequencing technologies
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../server_login/" class="md-nav__link">
        Setup
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../reproducibility/" class="md-nav__link">
        Reproducibility
      </a>
    </li>
  

            
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Quality control
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Quality control
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-outcomes" class="md-nav__link">
    Learning outcomes
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#material" class="md-nav__link">
    Material
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exercises" class="md-nav__link">
    Exercises
  </a>
  
    <nav class="md-nav" aria-label="Exercises">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#download-and-evaluate-an-e-coli-dataset" class="md-nav__link">
    Download and evaluate an E. coli dataset
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#run-fastqc" class="md-nav__link">
    Run fastqc
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#trim-the-reads" class="md-nav__link">
    Trim the reads
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5" type="checkbox" id="__nav_5" >
      
      
      
      
        <label class="md-nav__link" for="__nav_5">
          Day 2
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Day 2" data-md-level="1">
        <label class="md-nav__title" for="__nav_5">
          <span class="md-nav__icon md-icon"></span>
          Day 2
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../day2/read_alignment/" class="md-nav__link">
        Read alignment
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../day2/file_types/" class="md-nav__link">
        File types
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../day2/samtools/" class="md-nav__link">
        Samtools
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_6" type="checkbox" id="__nav_6" >
      
      
      
      
        <label class="md-nav__link" for="__nav_6">
          Day 3
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Day 3" data-md-level="1">
        <label class="md-nav__title" for="__nav_6">
          <span class="md-nav__icon md-icon"></span>
          Day 3
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../day3/igv_visualisation/" class="md-nav__link">
        IGV and visualisation
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../group_work/" class="md-nav__link">
        Group work
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#learning-outcomes" class="md-nav__link">
    Learning outcomes
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#material" class="md-nav__link">
    Material
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exercises" class="md-nav__link">
    Exercises
  </a>
  
    <nav class="md-nav" aria-label="Exercises">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#download-and-evaluate-an-e-coli-dataset" class="md-nav__link">
    Download and evaluate an E. coli dataset
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#run-fastqc" class="md-nav__link">
    Run fastqc
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#trim-the-reads" class="md-nav__link">
    Trim the reads
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  <a href="https://github.com/sib-swiss/NGS-introduction-training/edit/main/docs/day1/quality_control.md" title="Edit this page" class="md-content__button md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg>
  </a>


  <h1>Quality control</h1>

<h2 id="learning-outcomes">Learning outcomes</h2>
<p><strong>After having completed this chapter you will be able to:</strong></p>
<ul>
<li>Find information about a sequence run on the Sequence Read Archive</li>
<li>Run <code>fastqc</code> on sequence reads and interpret the results</li>
<li>Trim adapters and low quality bases using <code>fastp</code></li>
</ul>
<h2 id="material">Material</h2>
<iframe width="560" height="315" src="https://www.youtube.com/embed/iCOF6yaQFPE" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" allowfullscreen></iframe>

<p><a class="md-button" href="../../assets/pdf/quality_control.pdf"><span class="twemoji"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 384 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M64 0C28.7 0 0 28.7 0 64v384c0 35.3 28.7 64 64 64h256c35.3 0 64-28.7 64-64V160H256c-17.7 0-32-14.3-32-32V0H64zm192 0v128h128L256 0zM64 224h24c30.9 0 56 25.1 56 56s-25.1 56-56 56h-8v32c0 8.8-7.2 16-16 16s-16-7.2-16-16V240c0-8.8 7.2-16 16-16zm24 80c13.3 0 24-10.7 24-24s-10.7-24-24-24h-8v48h8zm72-64c0-8.8 7.2-16 16-16h24c26.5 0 48 21.5 48 48v64c0 26.5-21.5 48-48 48h-24c-8.8 0-16-7.2-16-16V240zm32 112h8c8.8 0 16-7.2 16-16v-64c0-8.8-7.2-16-16-16h-8v96zm96-128h48c8.8 0 16 7.2 16 16s-7.2 16-16 16h-32v32h32c8.8 0 16 7.2 16 16s-7.2 16-16 16h-32v48c0 8.8-7.2 16-16 16s-16-7.2-16-16V240c0-8.8 7.2-16 16-16z"/></svg></span> Download the presentation</a></p>
<ul>
<li><code>fastqc</code> command line <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/INSTALL.txt">documentation</a></li>
<li><code>cutadapt</code> <a href="https://cutadapt.readthedocs.io/en/stable/">manual</a></li>
<li>Unix command line <a href="https://www.ncbi.nlm.nih.gov/books/NBK179288/">E-utilities documentation</a></li>
</ul>
<h2 id="exercises">Exercises</h2>
<h3 id="download-and-evaluate-an-e-coli-dataset">Download and evaluate an E. coli dataset</h3>
<p>Check out the dataset at <a href="https://www.ncbi.nlm.nih.gov/sra/?term=SRR519926">SRA</a>.</p>
<p><strong>Exercise:</strong> Browse around the SRA entry and answer these questions:</p>
<p><strong>A.</strong> Is the dataset paired-end or single end?</p>
<p><strong>B.</strong> Which instrument was used for sequencing?</p>
<p><strong>C.</strong> What is the read length?</p>
<p><strong>D.</strong> How many reads do we have?</p>
<details class="done">
<summary>Answers</summary>
<p>A. paired-end</p>
<p>B. Illumina MiSeq</p>
<p>C. 2 x 251 bp</p>
<p>D. 400596</p>
</details>
<p>Now we will use some bioinformatics tools to do download reads and perform quality control. The tools are pre-installed in a conda environment called <code>ngs-tools</code>. Every time you open a new terminal, you will have to load the environment:</p>
<div class="highlight"><pre><span></span><code>conda<span class="w"> </span>activate<span class="w"> </span>ngs-tools
</code></pre></div>
<p>Make a directory <code>reads</code> in <code>~/workdir</code> and download the reads from the SRA database using <code>prefetch</code> and <code>fastq-dump</code> from <a href="https://ncbi.github.io/sra-tools/">SRA-Tools</a> into the <code>reads</code> directory. Use the code snippet below to create a scripts called <code>01_download_reads.sh</code>. Store it in <code>~/workdir/scripts/</code>, and run it.</p>
<div class="highlight"><span class="filename">01_download_reads.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nb">cd</span><span class="w"> </span>~/workdir
mkdir<span class="w"> </span>reads
<span class="nb">cd</span><span class="w"> </span>reads
prefetch<span class="w"> </span>SRR519926
fastq-dump<span class="w"> </span>--split-files<span class="w"> </span>SRR519926
</code></pre></div>
<p><strong>Exercise:</strong> Check whether the download was successful by counting the number of reads in the fastq files and compare it to the SRA entry.</p>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>A read in a fastq file consists of four lines (more on that at <a href="../../day2/file_types/">file types</a>). Use Google to figure out how to count the number of reads in a fastq file.</p>
</div>
<details class="done">
<summary>Answer</summary>
<p>e.g. from <a href="https://www.biostars.org/p/139006/">this</a> thread on Biostars:</p>
<div class="highlight"><pre><span></span><code><span class="c1">## forward read</span>
<span class="nb">echo</span><span class="w"> </span><span class="k">$(</span>cat<span class="w"> </span>SRR519926_1.fastq<span class="w"> </span><span class="p">|</span><span class="w"> </span>wc<span class="w"> </span>-l<span class="k">)</span>/4<span class="w"> </span><span class="p">|</span><span class="w"> </span>bc

<span class="c1">## reverse read</span>
<span class="nb">echo</span><span class="w"> </span><span class="k">$(</span>cat<span class="w"> </span>SRR519926_2.fastq<span class="w"> </span><span class="p">|</span><span class="w"> </span>wc<span class="w"> </span>-l<span class="k">)</span>/4<span class="w"> </span><span class="p">|</span><span class="w"> </span>bc
</code></pre></div>
</details>
<h3 id="run-fastqc">Run fastqc</h3>
<p><strong>Exercise:</strong> Create a script to run <code>fastqc</code> and call it <code>02_run_fastqc.sh</code>. After that, run it.</p>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p><code>fastqc</code> accepts multiple files as input, so you can use a <a href="https://en.wikipedia.org/wiki/Glob_(programming)">wildcard</a> to run <code>fastqc</code> on all the files in one line of code. Use it like this: <code>*.fastq</code>.  </p>
</div>
<details class="done">
<summary>Answer</summary>
<p>Your script <code>~/workdir/scripts/02_run_fastqc.sh</code> should look like:</p>
<div class="highlight"><span class="filename">02_run_fastqc.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>
<span class="nb">cd</span><span class="w"> </span>~/workdir/reads

fastqc<span class="w"> </span>*.fastq
</code></pre></div>
</details>
<p><strong>Exercise:</strong> Download the html files to your local computer, and view the results. How is the quality? Where are the problems?</p>
<div class="admonition info">
<p class="admonition-title">Downloading files</p>
<p>You can download files by right-click the file and after that select <strong>Download</strong>:</p>
<p><figure>
  <img src="../../assets/images/download_file.gif" width="500"/>
</figure></p>
</div>
<details class="done">
<summary>Answer</summary>
<p>There seems to be:</p>
<ul>
<li>Low quality towards the 3&rsquo; end (per base sequence quality)</li>
<li>Full sequence reads with low quality (per sequence quality scores)</li>
<li>Adapters in the sequences (adapter content)</li>
</ul>
<p>We can probably fix most of these issues by trimming.</p>
</details>
<h3 id="trim-the-reads">Trim the reads</h3>
<p>We will use <a href="https://github.com/OpenGene/fastp">fastp</a> for trimming adapters and low quality bases from our reads. The most used adapters for Illumina are TruSeq adapters, and <code>fastp</code> will use those by default. A reference for the adapter sequences can be found <a href="https://support.illumina.com/bulletins/2016/12/what-sequences-do-i-use-for-adapter-trimming.html">here</a>.</p>
<p><strong>Exercise:</strong> Check out the <a href="https://github.com/OpenGene/fastp">documentation of fastp</a>, and the option defaults by running <code>fastp --help</code>. 
- What is the default for the minimum base quality for a qualified base? ( option <code>--qualified_quality_phred</code>)
- What is the default for the maximum percentage of unqualified bases in a read? (option <code>--unqualified_percent_limit</code>)
- What is the default for the minimum required read length? (option <code>--length_required</code>)
- What happens if one read in the pair does not meet the required length after trimming? (it can be specified with the options <code>--unpaired1</code> and <code>--unpaired2</code>)</p>
<details class="done">
<summary>Answer</summary>
<ul>
<li>The minimum base quality is 15: <code>Default 15 means phred quality &gt;=Q15 is qualified. (int [=15])</code></li>
<li>The minimum required length is also 15: <code>reads shorter than length_required will be discarded, default is 15. (int [=15])</code></li>
<li>If one of the reads does not meet the required length, the pair is discarded if <code>--unpaired1</code> and/or <code>--unpaired2</code> are not specified: <code>for PE input, if read1 passed QC but read2 not, it will be written to unpaired1. Default is to discard it. (string [=])</code>. </li>
</ul>
</details>
<p><strong>Exercise:</strong> Complete the script below called <code>03_trim_reads.sh</code> (replace everything in between brackets <code>[]</code>) to run <code>fastp</code> to trim the data.  The quality of our dataset is not great, so we will overwrite the defaults.  Use a a minimum qualified base quality of 10, set the maximum percentage of unqalified bases to 80% and a minimum read length of 25. Note that a new directory called <code>~/workdir/results/trimmed/</code> is created to write the trimmed reads.</p>
<div class="highlight"><span class="filename">03_trim_reads.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nv">TRIMMED_DIR</span><span class="o">=</span>~/workdir/results/trimmed
<span class="nv">READS_DIR</span><span class="o">=</span>~/workdir/reads

mkdir<span class="w"> </span>-p<span class="w"> </span><span class="nv">$TRIMMED_DIR</span>

<span class="nb">cd</span><span class="w"> </span><span class="nv">$TRIMMED_DIR</span>

fastp<span class="w"> </span><span class="se">\</span>
-i<span class="w"> </span><span class="nv">$READS_DIR</span>/SRR519926_1.fastq<span class="w"> </span><span class="se">\</span>
-I<span class="w"> </span><span class="nv">$READS_DIR</span>/SRR519926_2.fastq<span class="w"> </span><span class="se">\</span>
-o<span class="w"> </span><span class="nv">$TRIMMED_DIR</span>/trimmed_SRR519926_1.fastq<span class="w"> </span><span class="se">\</span>
-O<span class="w"> </span><span class="nv">$TRIMMED_DIR</span>/trimmed_SRR519926_2.fastq<span class="w"> </span><span class="se">\</span>
<span class="o">[</span>QUALIFIED<span class="w"> </span>BASE<span class="w"> </span>THRESHOLD<span class="o">]</span><span class="w"> </span><span class="se">\</span>
<span class="o">[</span>MINIMUM<span class="w"> </span>LENGTH<span class="w"> </span>THRESHOLD<span class="o">]</span><span class="w"> </span><span class="se">\</span>
<span class="o">[</span>UNQUALIFIED<span class="w"> </span>PERCENTAGE<span class="w"> </span>LIMIT<span class="o">]</span><span class="w"> </span><span class="se">\</span>
--cut_front<span class="w"> </span><span class="se">\</span>
--cut_tail<span class="w"> </span><span class="se">\</span>
--detect_adapter_for_pe
</code></pre></div>
<div class="admonition note">
<p class="admonition-title">Additional options</p>
<p>Note that we have set the options <code>--cut_front</code> and <code>--cut_tail</code> that will ensure low quality bases are trimmed in a sliding window from both the 5&rsquo; and 3&rsquo; ends. Also <code>--detect_adapter_for_pe</code> is set, which ensures that adapters are detected automatically for both R1 and R2. </p>
</div>
<details class="done">
<summary>Answer</summary>
<p>Your script (<code>~/workdir/scripts/03_trim_reads.sh</code>) should look like this:</p>
<div class="highlight"><span class="filename">03_trim_reads.sh</span><pre><span></span><code><span class="ch">#!/usr/bin/env bash</span>

<span class="nv">TRIMMED_DIR</span><span class="o">=</span>~/workdir/results/trimmed
<span class="nv">READS_DIR</span><span class="o">=</span>~/workdir/reads

mkdir<span class="w"> </span>-p<span class="w"> </span><span class="nv">$TRIMMED_DIR</span>

<span class="nb">cd</span><span class="w"> </span><span class="nv">$TRIMMED_DIR</span>

fastp<span class="w"> </span><span class="se">\</span>
-i<span class="w"> </span><span class="nv">$READS_DIR</span>/SRR519926_1.fastq<span class="w"> </span><span class="se">\</span>
-I<span class="w"> </span><span class="nv">$READS_DIR</span>/SRR519926_2.fastq<span class="w"> </span><span class="se">\</span>
-o<span class="w"> </span><span class="nv">$TRIMMED_DIR</span>/trimmed_SRR519926_1.fastq<span class="w"> </span><span class="se">\</span>
-O<span class="w"> </span><span class="nv">$TRIMMED_DIR</span>/trimmed_SRR519926_2.fastq<span class="w"> </span><span class="se">\</span>
--qualified_quality_phred<span class="w"> </span><span class="m">10</span><span class="w"> </span><span class="se">\</span>
--length_required<span class="w"> </span><span class="m">25</span><span class="w"> </span><span class="se">\</span>
--unqualified_percent_limit<span class="w"> </span><span class="m">80</span><span class="w"> </span><span class="se">\</span>
--cut_front<span class="w"> </span><span class="se">\</span>
--cut_tail<span class="w"> </span><span class="se">\</span>
--detect_adapter_for_pe
</code></pre></div>
</details>
<div class="admonition note">
<p class="admonition-title">The use of <code>\</code></p>
<p>In the script above you see that we&rsquo;re using <code>\</code> at the end of many lines. We use it to tell bash to ignore the newlines. If we would not do it, the <code>fastp</code> command would become a very long line, and the script would become very difficult to read. It is in general good practice to put every option of a long command on a newline in your script and use <code>\</code> to ignore the newlines when executing.</p>
</div>
<p><strong>Exercise:</strong> Check out the report in <code>fastp.thml</code>. </p>
<ul>
<li>Has the quality improved?</li>
<li>How many reads do we have left?</li>
<li><em>Bonus</em>: Although there were adapters in R2 according to <code>fastqc</code>,  <code>fastp</code> has trouble finding adapters in R2. Also, after running <code>fastp</code> there doesn&rsquo;t seem to be much adapter left (you can double check by running <code>fastqc</code> on <code>trimmed_SRR519926_2.fastq</code>). How could that be? </li>
</ul>
<details class="done">
<summary>Answers</summary>
<ul>
<li>Yes, low quality 3&rsquo; end, per sequence quality and adapter sequences have improved. Also the percentages &gt;20 and &gt;30 are higher. </li>
<li>624724 reads, so 312362 pairs (78.0%)</li>
<li>The 3&rsquo; end of R2 has very low quality on average, this means that trimming for low quality removes almost all bases from the original 3&rsquo; end, including any adapter.  </li>
</ul>
</details>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
    <nav class="md-footer__inner md-grid" aria-label="Footer" >
      
        
        <a href="../reproducibility/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Reproducibility" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Reproducibility
            </div>
          </div>
        </a>
      
      
        
        <a href="../../day2/read_alignment/" class="md-footer__link md-footer__link--next" aria-label="Next: Read alignment" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Read alignment
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "../..", "features": [], "search": "../../assets/javascripts/workers/search.16e2a7d4.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "version": {"provider": "mike"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.8492ddcf.min.js"></script>
      
    
  </body>
</html>